File: products/Infra/Rabbit/Publishers/Publisher.cs
using System;
using System.Text;
using Infra.Interfaces.Publishers;
using Newtonsoft.Json;
using Newtonsoft.Json.Serialization;
using RabbitMQ.Client;

namespace Infra.Rabbit.Publishers
{
    public abstract class Publisher<T> : IPublisher<T>
    {
        private readonly string _exchange;
        private readonly IModel _channel;

        protected Publisher(String exchange, IModel channel)
        {
            _exchange = exchange;
            _channel = channel;

            _channel.ExchangeDeclare(exchange, "fanout", false, false, null);
        }

        public void Publish(T data)
        {
            _channel.ExchangeDeclare(_exchange, "fanout", false, false, null);

            string message = JsonConvert.SerializeObject(data, new JsonSerializerSettings
            {
                ContractResolver = new CamelCasePropertyNamesContractResolver()
            });

            var body = Encoding.UTF8.GetBytes(message);

            _channel.BasicPublish(_exchange, "", null, body);
        }
    }
}

File: buyer/src/main/java/com/madness/microservice/infra/rabbitmq/publishers/Publisher.java
package com.madness.microservice.infra.rabbitmq.publishers;

import java.io.IOException;

import com.google.gson.Gson;
import com.madness.microservice.infra.gson.GsonSerializer;
import com.madness.microservice.infra.rabbitmq.Exchanges;
import com.madness.microservice.infra.rabbitmq.RabbitMqConnection;

public abstract class Publisher<T> {

  private RabbitMqConnection _rabbitMq;
  private Gson _gson;

  protected abstract Exchanges exchange();

  protected Publisher() {
  }

  public Publisher(RabbitMqConnection rabbitMq, GsonSerializer serializer) {
    _rabbitMq = rabbitMq;
    _gson = serializer.gson;
  }

  public void publish(T data) throws IOException {

    var bytes = _gson.toJson(data).getBytes();

    var channel = _rabbitMq.connection().createChannel();

    channel.exchangeDeclare(exchange().getValue(), "fanout", false);
    
    channel.basicPublish(exchange().getValue(), "", null, bytes);
  }
}